### PR TITLE
Handling idempotent commands in EventSourcedBehavior

### PR DESCRIPTION
**Pull request by ptrdom**
I was working on this POC for idempotent command handling in `EventSourcedBehavior`, and thought it's in a good state to be presented for consideration before starting to work on various tasks related to full integrating this as a feature for future versions of Akka. 

Basically this adds capability to assigning an ID - idempotency key - for a specific command, and said command will only be successfully processed once. Idempotency keys are saved to journal, either atomically with events produced by command or separately if there are none. Here are some diagrams for clarification.

idempotency-solution-abstract

A simple class hierarchy was added for implementation of idempotent commands and their replies. Idea is that domain logic in command handler returns `IdempotenceSuccess`, and in case of duplicate processing `IdempotenceFailure` gets returned with state of entity - some useful data for caller to continue it's domain logic even in case of failure. Didn't have good ideas how to add better type safety here.

idempotency-command-reply-model

There were also a few changes to `Running` states of `EventSourcedBehavior` - idempotency key LRU cache in `HandlingCommands` and two new states for idempotence related interactions with event journal.

idempotency-solution-eventsourcedbehavior

I was also working on event journal implementation for Apache Cassandra. Idempotency keys obviously must use same journal as events, because they are all persisted together atomically. Journal implementation is a separate topic of course, but the idea was to partition events and idempotency keys together, and boy oh boy that requires some heavy changes. I assume JDBC journal implementation should be a piece of cake, because there's no partitioning control there. If this gets accepted I could then continue the work on Apache Cassandra journal, and I'd need some guidance to help integrate it smoothly.

To sum this up, I think this functionality fits quite nicely with Akka and provides a functionality I'd personally love to have in event sourcing, as it helps a ton when handing interactions with command producers that in best case provide at-least-once delivery. There are no breaking changes to public interfaces, but binary compatibility is a tricky topic. 

What do you think about this?

### CODE REVIEW

**Comment by renatocaval** [comment]
@dope9967, is that something that should be part of the domain logic of a `EventSourcedBehavior`? 

I mean, if I want to have this kind of functionality, I could add it to my  domain and make sure that all events are persisted with the version key. I would then keep the version key on my actor state and use it as a validation filter for incoming commands. 

That wouldn't require changing any of the existing plugins. 

Of course, that would require you to implement it over and over again for each project, but this can be solved by having some utility classes that adds that functionality.

**Comment by patriknw** [comment]
@dope9967 Thanks for engaging and contributing to Akka. This looks like a massive change to the internals of EventSourceBehavior, and also the journals. I'm not sure we want to add this feature on that layer. How would it look like if it was implemented by the application instead and documented as a pattern?

**Comment by dope9967** [comment]
@renatocaval if I understand you example correctly, that wouldn't work well if one entity reacts to events from multiple entities and/or receives commands from external clients, then some coordination for version keys would be required, and I cannot figure out right away how that wouldn't get too complicated. I could have misunderstood your example, could you maybe elaborate a bit?

@patriknw the main idea of this feature was to separate this from domain logic. It's possible to implement this now in it, but that causes several issues:
* Fundamentally it's a separate concern, so it would be ideal to solve it in some higher abstraction - utility, library implementation - so it's easily reusable.
* Keeping this in domain logic means idempotency keys have to be aggregated into entity's state and also be stored in it's events. That adds an ever increasing memory/disk footprint to handle the entity's actor and store/transport it's snapshots. Also this goes back to the first point of mixing lower level idempotence logic with domain logic, and no matter how you go about it, it's still kind of unwieldy.

Proposed implementation of idempotence functionality for `EventSourcedBehavior` doesn't really affect systems not using it - if command is not an `IdempotentCommand`, then execution is basically the same as before, and the most complicated part is journal implementations, so maybe journal interfaces could be implemented in such way that would allow idempotence to be optional journal feature? Then it wouldn't require journal developers to jump right on it.

Overall this is a pretty big change, I can understand the reluctance to move forward with it, especially if the value does not appear great. Maybe I have some wrong idea about design of more complicated interactions, but I've noticed that eventual consistence is enough of an headache and with at-least-once delivery thrown in, in certain scenarios it can get hard to reason about what's going on. If you'll still consider to not go forward with this feature, I'll completely understand it, though personally for me it would be very cool to work on this.

**Comment by renatocaval** [comment]
I was referring to something similar to optimistic locking. Whenever you send a command, you have to know each version the entity is. 

In your initial description, you wrote:

> Basically this adds capability to assigning an ID - idempotency key - for a specific command, and said command will only be successfully processed once. Idempotency keys are saved to journal, either atomically with events produced by command or separately if there are none. Here are some diagrams for clarification.

The idempotency key needs to increase monotonically otherwise you need to keep track of all keys that were ever processed. Searching for a non-monotonically increased key will incur in perf penalties. 

Moreover, saving the key and the events can't be done atomically (in case of Cassandra) and that would remove any the idempotency guarantees. The key needs to be part of the event if you want to save it atomically. And as a consequence you need to apply the key to the model, the model must know the latest key.

With that in mind, and to my understand, we can only have idempotency based on versioning. The model should keep track in which version it is and reject or ignore commands that send a previous version number.

**Comment by dope9967** [comment]
@renatocaval thank you for elaborating on your example - now I understand the cases where such implementation is useful. I'll elaborate a bit on idempotency key storage implementation for Cassandra.

Are you sure that saving keys and events can't be done atomically? Because info that I've found in docs and forums ([link1](https://docs.datastax.com/en/cql-oss/3.3/cql/cql_using/useBatch.html), [link2](https://stackoverflow.com/questions/49356986/cassandra-batch-statement-multiple-tables)) does not indicate such limitation. Only issue is number of partitions involved in batch and it's size - performance degrades quite quickly for multiple partition writes and with increasing batch size. Having that in mind, I've investigated if partitions of different tables can be stored in same partition and although I couldn't find any info in docs, I did find several forum posts confirming that if multiple tables are stored in same keyspace, use same partition keys and replication strategies, then their data will be stored in same nodes ([link3](https://community.datastax.com/questions/4418/cassandra-what-is-guaranteed-with-respect-to-the-t.html), [link4](https://stackoverflow.com/questions/61982652/cassandra-what-is-guaranteed-with-respect-to-the-tables), [link5](https://stackoverflow.com/questions/34294830/how-to-keep-2-cassandra-tables-within-same-partition)). Moving forward with these findings, I've concluded that following table structure could be implemented:

1. TABLE idempotency_keys_search 
  `PRIMARY KEY ((persistence_id, partition_nr), idempotency_key)`
2. TABLE idempotency_keys_cache 
  `PRIMARY KEY ((persistence_id, partition_nr), sequence_nr, idempotency_key) WITH CLUSTERING ORDER BY (sequence_nr DESC)
`

First table would be used for search, and, as you've indicated, nothing much can be done for performance there since any idempotency key can be potentially queried, only potential speedup could be parallel querying.  Second table would be used for replaying cache upon actor activation and finding out highest sequence number of idempotency keys. Knowing highest sequence numbers of idempotency keys and events allows storing all the data of same batch in same partition - `maxPartitionNr = (highestEventSequenceNr + (highestIdempotencyKeySequenceNr * 2) - 1) / partitionSize` - but means that event partition determinism is lost - `eventPartitionNr != (eventSequenceNr - 1) / partitionSize` - and that's the toughest issue that I've encountered with the implementation, because much of the code depends on it. This implementation also allows control of partition size to be retained.

I can understand why moving forward with this implementation wouldn't be considered because of unavoidable negative performance impact in certain cases, but if it's still deemed to be a useful feature, and if pros/cons were clearly documented and the feature is made optional, then developers themselves could decide if they a willing to use it?

What do you think about this implementation?

**Comment by patriknw** [comment]
We are not going to introduce lightweight (paxos) transactions in Akka Persistence Cassandra, if that is what you refer to.
It's possible to persist two events atomically when using the Effect.persist(event1, event2).

I'd rather see this as a feature on top of EventSourcedBehavior, in a separate library maintained by you. Alternatively, as a documented pattern (sample), which doesn't aim for separating this concern from the application logic. A library can have some utilities to maintain the state and create Persist effects, but it will probably not be completely invisible to the user.

**Comment by dope9967** [comment]
Got it, all good. I'm not sure if this functionality has enough value with alternative implementation, I'll need to think this through more. Thank you for your time to discuss this.
